KAFKA-5925: Adding records deletion operation to the new Admin Client API

ppatierno opened the pull request:
This is the PR related to the [KIP-204](https://cwiki.apache.org/confluence/display/KAFKA/KIP-204+%3A+Adding+records+deletion+operation+to+the+new+Admin+Client+API) in order to add the `deleteRecords` operation to the new Admin Client (it's already available in the "legacy" one).
Other than that, unit test and integration tests are added as well (such integration tests come from the "legacy" integration tests in order to test the new addition in the same way as the "legacy" one).

ppatierno committed:
Started to add skeleton about delete records operation on Admin Client

ppatierno committed:
Added first skeleton for deleteRecords method in the Admin Client

ppatierno committed:
Completed the DeleteRecordsResult
Added handle response logic for the deleteRecords operation

ppatierno committed:
Fixed wrong checking on no error condition

ppatierno committed:
Added a unit test on Admin Client delete records operation

ppatierno committed:
Updated inheritance of AdminClientIntegrationTest from IntegrationTestHarness for consumer/producer support (delete records tests to add)

ppatierno committed:
Added DeleteRecords class for wrapping long offset

ppatierno committed:
Updated AdminClient interface and implementation with new DeleteRecords class

ppatierno committed:
Renamed DeleteRecords to DeleteRecordsTarget

ppatierno committed:
Modified getting metadata strategy using a nested Call approach (metadata then delete records request)
Added integration tests as "legacy" admin client tests
Added a unit test

ppatierno committed:
Minor fix on creating Map instances

ppatierno committed:
Renamed DeleteRecordsTarget to RecordsToDelete with related methods
Fixed not working unit test for records deletion

ppatierno committed:
Minor fixes on unit and integration tests

ppatierno committed:
Fixed some style errors after checking

ppatierno committed:
Fixed conflict with trunk

ppatierno committed:
Merge branch 'trunk' into kafka-5925

ppatierno committed:
Fixed conflict on generateConfigs method

tedyu commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
+        }
+
+        // preparing topics list for asking metadata about them
+        final List<String> topics = new ArrayList<>();
```
Should Set be used here (since there may be duplicate topics from the loop below) ?

ppatierno replied:
You are right. I have just fixed that. Thanks !

ppatierno committed:
Fixed topics from List to Set

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -508,4 +509,30 @@ public CreatePartitionsResult createPartitions(Map<String, NewPartitions> newPar
                                                             CreatePartitionsOptions options);
 
+    /**
+     * Delete records from topic partitions starting at the provided offsets.
```
nit: `starting at the offsets` may be a bit confusing to users, as it may be interpreted as "from given offset to leo" whereas we are actually deleting "from lso to given offset (exclusive)". The javadoc on the old client may be more clarifying:

```
.. whose offset is smaller than the given offset of the corresponding partition
```

ppatierno replied:
Totally agree.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteRecordsResult.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Map;
+
+@InterfaceStability.Evolving
+public class DeleteRecordsResult {
```
I think we need to add the remaining javadocs in this class since it will be a public interface. For example, the `Long` in the `KafkaFuture` result, my guess it that would be low watermark?

ppatierno replied:
I use to comment too much the code, I don't know why I forgot this kind of javadoc information.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/RecordsToDelete.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * The offset before which all records will be deleted
+     */
+    public long offset() {
```
nit: rename to `beforeOffset`?

ppatierno replied:
Done

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/RecordsToDelete.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    public String toString() {
+        return "(offset = " + offset + ")";
```
Ditto above.

ppatierno replied:
Done

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
+        }
+
+        final long nowMetadata = time.milliseconds();
+        final long deadline = calcDeadlineMs(nowMetadata, options.timeoutMs());
```
Hmm.. we already have a `metadata` object that is keeping updated by the `AdminClientRunnable`, can we just call `metadata.fetch()` to get the current cluster information? Then in line 1918 if we do not have the current leader we can still return `LEADER_NOT_AVAILABLE` to let the caller retry as it is a retryable error code.

cmccabe replied:
I don't think that will work, because the metadata object that is maintained in AdminClientRunnable does not fetch information about any topics.  In general the way the Metadata object works is a very poor fit for AdminClient-- it was designed for producers and consumers, where you have long-running subscriptions and so forth.  It is much simpler and better just to make the metadata call here-- we can always optimize this later with a cache if needed.

ppatierno replied:
@guozhangwang it was the way I started with the implementation but then I noticed what @cmccabe is saying (no topics information fetched). My first way was to do a metadata request with topics in a similar way the producer does before using the `AdminClientRunnable` for the delete records call. I didn't like it too much (even because it was synchronous) so I switched to use the `AdminClientRunnable` for both (metadata and delete requests) leveraging on what is already working well today.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
+                // completing futures for topics with errors
+                for (Map.Entry<String, Errors> topicError: response.errors().entrySet()) {
+                    KafkaFutureImpl<Long> future = futures.get(topicError.getKey());
+                    future.completeExceptionally(topicError.getValue().exception());
```
See my comment above, I'm not sure if we should expose an error that is returned from the metadata request to users' delete record response.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                completeAllExceptionally(futures.values(), throwable);
```
Ditto here. I think we should consider getting rid of the metadata request and exposing any exceptions from this request to the user's expected delete record response, instead we just rely on the whatever the current metadata (up-to-date or not) and if there is no leader known we set the future exception immediately.

cmccabe replied:
Please, let's not.  The other functions in AdminClient do not rely on metadata caching-- they use the latest metadata that is available.  Deleting records shouldn't be a common operation.  If it is, we can have a metadata cache with a configurable expiration time.  I think it's also really bad to set an exception based on possibly stale information.  You give the user no way out if the cache is stale (besides creating an entirely new admin client object, I suppose).

cmccabe replied:
As per our offline discussion, we'll leave it like this for now

guozhangwang reviewed (commented):
@ppatierno Thanks for the patch. I made a pass over the non-testing classes. Left one meta comment on metadata update, other comments are minor.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteRecordsResult.java:
```diff
@@ -0,0 +1,49 @@
+     * Return a map from topic partition to futures which can be used to check the status of
+     * individual deletions.
+     */
+    public Map<TopicPartition, KafkaFuture<Long>> values() {
```
`values` is a very generic name.  How about calling this `lowWatermarks`?

ppatierno replied:
@cmccabe I agree with you but I'm following what seems to be the common pattern even for other results in the Admin Client. For example, `DeleteAclsResult`, `DescribeConfigsResult`, `DescribeReplicaLogDirsResult`, `DescribeTopicsResult`. It seems that only `ListTopicsResult` and `DescribeClusterResult` are using meaningful names. I'm happy to change here from `values` to `lowWatermarks` but what do you think if I create a JIRA and a corresponding PR for giving meaningful names to the corresponding methods for above classes as well ?

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
+
+            @Override
+            AbstractRequest.Builder createRequest(int timeoutMs) {
+                return new MetadataRequest.Builder(new ArrayList<>(topics), false);
```
I don't think this will work, will it?  You should specify the topics that you want metadata about.  If you specify an empty list, no topic info will be returned.

guozhangwang replied:
With an empty list it means "return me every topic you know".

cmccabe replied:
null means "return me every topic you know".  The empty list means no topics.  (This changed in a previous AK version)

ppatierno replied:
> I don't think this will work, will it? You should specify the topics that you want metadata about. If you specify an empty list, no topic info will be returned.

@cmccabe what do you mean ? the `topics` list is built few lines before starting from the map provided. I'm pretty sure you noticed that so I'm missing your point here, sorry :-(

cmccabe replied:
Sorry, my earlier comment was related to an earlier version of the patch.

guozhangwang commented:
@cmccabe @ppatierno Here are my rationale to not enforcing a metadata refresh before send the delete records request:

* First, about metadata freshness: even if you force refreshing the metadata every time before you are trying to send any request, you are still vulnerable to stale information since the cluster metadata could still change while you are getting the metadata response. **On the other hand**, in practice metadata does not change very often. So we are actually trying to get rid of like 9% of the race conditions by always enforcing a metadata refresh, while for other 90% it just works fine, plus we can never make it 100% due to the 1% scenario I mentioned above. To me this is not a good trade-off.

* Second, about the latency: delete records request incurs at least three round trips to get a response: one between client and the leader, one between leader and follower to let follower learn the new starting offset, and one between the leader / follower again to let leader know that all followers have set their starting offsets. Enforcing a metadata refresh before this request would just further increase its latency. Although with the async KafkaAdminClient we are not necessarily going to be blocking, it would still reduce its usefulness with larger latencies.

* Finally, about the user-facing interface: I understand that you may consider returning an error that could be avoided if we have fresh to users sounds bad, but like I said we will not be able to get 100% safety ever, and that's why we designed the error code LEADER_NOT_AVAILABLE as retriable error code in the protocol to let users retry. And users are expected to handle such cases anyways. So I think this does not actually incur much burden on them.

What we did in the producer / consumer client, is to only refresh metadata based on period (metadata.age is default to 5min) AND upon failures like "I'm not the controller / coordinator / leader", which are all retriable. I think we should do the same in AdminClient (currently it seems we only try to get the metadata once), to request a metadata refresh while returning the retriable error code to users.

guozhangwang commented:
I had some sync with @cmccabe offline on whether to apply this metadata caching optimization in this PR, and we have agreed on not doing this yet. So I think we can enforcing a metadata request on each call.

Aside from another meta comment on updating the javadoc on possible error codes / exceptions from the `Future<..>` I do not have any additional comments. Note that we may return both `LEADER_NOT_AVAILABLE` (if the metadata response does not contain that topic) and `NOT_LEADER_FOR_PARTITION` (if the metadata get stale while we are grouping the requests and send them). On these cases users are expected to retry.

@ppatierno Could you rebase also?

ppatierno committed:
Added more Javadocs and made some re-naming

ppatierno committed:
Merge branch 'trunk' into kafka-5925

ppatierno commented:
@guozhangwang @cmccabe I have made all the requested changes and rebased.

guozhangwang commented:
retest this please

asfgit commented:
FAILURE
 No test results found.
--none--

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -414,6 +419,66 @@ public void testCreatePartitions() throws Exception {
+
+        HashMap<Integer, Node> nodes = new HashMap<>();
+        nodes.put(0, new Node(0, "localhost", 8121));
+        List<PartitionInfo> partitionInfos = new ArrayList<>();
```
Could we update the `mockClientEnv` function to include these leader information? I understand that only this test and the missing test for `listTopics` (we'd probably add that asap cc @cmccabe ), but maybe not hurt to add that into the general metadata.

ppatierno replied:
@guozhangwang if I understood your suggestion it sounds to me to change the `mockClientEnv` signature in this way : 

`mockClientEnv(Collection<PartitionInfo> partitions, String... configVals)`

having all the current tests which don't need it, passing just `Collections.<PartitionInfo>emptySet()` while the current delete records test builds the collection and then pass it to the method. Is this what were you thinking ?

ppatierno replied:
@guozhangwang if I understood your suggestion it sounds to me to change the `mockClientEnv` signature in this way :

`mockClientEnv(Collection<PartitionInfo> partitions, String... configVals)`

having all the current tests, which don't need it, passing just `Collections.<PartitionInfo>emptySet()` while the current delete records test builds the collection and then pass it to the method. Is this what were you thinking ?

guozhangwang replied:
Yes that was my suggestion.

ppatierno replied:
I have taken a look to the other tests. All of them use just the `mockClientEnv` method. Only two tests build the cluster from scratch : `testDeleteRecords` and `testHandleTimeout`.
I can change the `mockClientEnv` for using it in the `testDeleteRecords` but we still have the problem for the `testHandleTimeout`. Maybe we should leave the `mockClientEnv` as it is because it works for the most of the tests and handling specific use cases as we do today ?

guozhangwang replied:
Yeah I think this is a subjective opinion. I'm fine with it staying as is.

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -414,6 +419,66 @@ public void testCreatePartitions() throws Exception {
+                                    Collections.singletonList(nodes.get(0)), Collections.singletonList(nodes.get(0)), Collections.<Node>emptyList()))));
+
+
+            env.kafkaClient().prepareResponse(new MetadataResponse(cluster.nodes(), cluster.clusterResource().clusterId(), cluster.controller().id(), t));
```
Maybe we could also test the case when `MetadataResponse` does not include the topic, and when the later `DeleteRecordsResponse` returns an error code that claim not the leader.

More generally speaking, we could have for error code in the response of five different partitions, with `OFFSET_OUT_OF_RANGE`, `UNKNOWN_TOPIC_OR_PARTITION`, `LEADER_NOT_AVAILABLE`, `NOT_LEADER_FOR_PARTITION` and `NONE`, and test that these error codes are correctly returned when the specific scenarios happens.

ppatierno replied:
Ok. I have updated the unit test using the same "my_topic" topic but 5 different partitions for simulating different errors (or success) to check.

guozhangwang commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -75,11 +77,12 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
 
   override def generateConfigs = {
-    val cfgs = TestUtils.createBrokerConfigs(brokerCount, zkConnect, interBrokerSecurityProtocol = Some(securityProtocol),
+    val cfgs = TestUtils.createBrokerConfigs(serverCount, zkConnect, interBrokerSecurityProtocol = Some(securityProtocol),
```
In line 98 below, do we still need to add `cfgs.foreach(_ ++= serverConfig)`?

ppatierno replied:
Currently it's not needed even because the `serverConfig` is just empty. Maybe in the future some other tests will be added to this class configuring more server configs ?

guozhangwang replied:
Makes sense, let's leave it here then.

guozhangwang commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -698,6 +701,202 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
     }
   }
 
+  val topic = "topic"
```
nit: move these to the top of the class.

ppatierno replied:
Done

guozhangwang commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -698,6 +701,202 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+  }
+
+  @Test
+  def testConsumeAfterDeleteRecords(): Unit = {
```
Personally I feel the previous test can be merged into this one: after seekToBeginning and check that position is 5, seek(1) and then consume.

ppatierno replied:
It's not so clear to me the sequence you are proposing here in relation to what the actual test does. Can you elaborate more please ?

guozhangwang replied:
The first test checks that after the deletion, lw has updated and hence seek-to-beginning will respect this new lw.

The second test checks that after the deletion, lw has updated and hence seek-to-offset will respect this new lw.

I'm thinking maybe we can consolidate thest two test to 1) call delete records, 2) seek-to-beginning and check position, 3) seek-to-offset and check the position.

BTW I also think we do not need consumer poll checking that the expected num. messages are returned, since "fetching from the position to leo returns these many num. records" is covered by the consumer unit test already, so this sounds a bit overlap to me.

ppatierno replied:
Ah Ok. Got it ! just done ;)

guozhangwang commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -698,6 +701,202 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+  }
+
+  @Test
+  def testDeleteRecordsWithException(): Unit = {
```
I feel this test can actually be moved into the unit test class, as it is not testing the interaction of more than two modules actually. In fact we can just mock the metadata and prepare the responses, and then check the returned exception.

ppatierno replied:
Actually the unit test already had an "offset out of range" exception check. Now I have also added a "leader not available" exception check as well. I have removed this integration test too.

guozhangwang reviewed (commented):
@ppatierno I made another pass over the test classes. Overall looks good to me. Will wait for the java7/8 jenkins jobs to succeed.

We also need to wait for KIP-204 to pass the vote.

asfgit commented:
FAILURE
 7955 tests run, 5 skipped, 1 failed.
--none--

guozhangwang commented:
I saw this test failure from Jenkins: https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/9267/testReport/junit/kafka.api/SaslSslAdminClientIntegrationTest/testLogStartOffsetCheckpoint/

Is it related to your change?

asfgit commented:
SUCCESS 
 8063 tests run, 5 skipped, 0 failed.
--none--

ppatierno committed:
Fixed bug on delete records metadata request
Moved testDeleteRecordsWithException from integration to unit test
Changed testDeleteRecords unit test using only one topic with five partitions for testing errors (and success)
Moved topic partition declaration at the top of AdminClientIntegrationTest class

ppatierno commented:
@guozhangwang regarding the failed test I see a page to 404 error :-(

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

ppatierno committed:
Fixed bug on using keySet

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 7953 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
SUCCESS 
 8061 tests run, 5 skipped, 0 failed.
--none--

ppatierno committed:
Added catching for the retry logic in the testLogStartOffsetCheckpoint test

guozhangwang commented:
retest this please

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
SUCCESS 
 8062 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8062 tests run, 5 skipped, 0 failed.
--none--

ppatierno committed:
Consolidate integration tests about seeking after records deletion

asfgit commented:
FAILURE
 5034 tests run, 1 skipped, 1 failed.
--none--

asfgit commented:
SUCCESS 
 8062 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8062 tests run, 5 skipped, 0 failed.
--none--

ppatierno committed:
Renaming partitionsAndOffsets parameter to recordsToDelete

asfgit commented:
FAILURE
 7181 tests run, 1 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 7968 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 7968 tests run, 5 skipped, 1 failed.
--none--

ppatierno committed:
Added DeleteRecords class as wrapper as result of records deletion
Updated Admin Client and related unit tests with above class usage

asfgit commented:
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
FAILURE
 7968 tests run, 5 skipped, 1 failed.
--none--

ppatierno commented:
retest this please

asfgit commented:
FAILURE
 7968 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteRecords.java:
```diff
@@ -0,0 +1,47 @@
+     *
+     * @param lowWatermark  "low watermark" for the topic partition on which the deletion was executed
+     */
+    public DeleteRecords(long lowWatermark) {
```
Does it makes sense to make this a factory function like:
```
public static DeleteRecords beforeLowWatermark(long lowWatermark) { ... }
```

ppatierno replied:
I think that having a factory method makes sense for `RecordsToDelete` because it's something that the user uses. We added the factory method for having a clearer API for the users. The `DeleteRecords` class is instantiated inside the Admin Client implementation (other than just the unit tests) so I don't think it makes more sense having such a method.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,104 @@ void handleFailure(Throwable throwable) {
+                        }
+                    }
+                }
+
```
I might be missing something here, but aren't we still going to make deleteRecords calls in this loop even for topics that had errors in the loop above?

ppatierno replied:
No, you are absolutely right. Just fixed.

guozhangwang replied:
Thanks for the catch!

cmccabe commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -75,11 +81,12 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
 
-  val brokerCount = 3
-  lazy val serverConfig = new Properties
+  val serverCount = 3
```
Is there a reason to change brokerCount to serverCount?  brokerCount seems clearer to me?

ppatierno replied:
It's because now the `AdminClientIntegrationTest` class is derived from `IntegrationTestHarness` where we have `serverCount` alredy defined. We should change `serverCount` to `brokerCount` in such base class. It means to change the name for other class derived from `IntegrationTestHarness` (i.e. `MetricsTest` and so on). Wdyt ?

guozhangwang replied:
I'm +1 for the improvement on reusing the base class's `serverCount`, as for its name I'm fine with either one..

ppatierno replied:
@cmccabe wdyt about that ?

ppatierno committed:
Renamed DeteleRecords to DeletedRecords

ppatierno committed:
Fixed avoiding to send deletion request for topics with errors

asfgit commented:
FAILURE
 7968 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 7968 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 8076 tests run, 5 skipped, 2 failed.
--none--

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,109 @@ void handleFailure(Throwable throwable) {
+                for (Map.Entry<TopicPartition, RecordsToDelete> entry: recordsToDelete.entrySet()) {
+
+                    // avoiding to send deletion request for topics with errors
+                    if (!response.errors().containsKey(entry.getKey().topic())) {
```
nit: Note that `response.errors()` will reconstruct the map from `topicMetadata` on every call. Could we just keep a reference of it in line 1909 above and reuse the map here?

ppatierno replied:
it's not a nit ... it's a must ;) thanks !

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,109 @@ void handleFailure(Throwable throwable) {
+                    // avoiding to send deletion request for topics with errors
+                    if (!response.errors().containsKey(entry.getKey().topic())) {
+
+                        Node node = response.cluster().leaderFor(entry.getKey());
```
nit: ditto here, `cluster()` will reconstruct a new object on each call.

ppatierno replied:
it's not a nit ... it's a must ;) thanks !

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1871,4 +1873,109 @@ void handleFailure(Throwable throwable) {
+                    if (!response.errors().containsKey(entry.getKey().topic())) {
+
+                        Node node = response.cluster().leaderFor(entry.getKey());
+                        if (node != null) {
```
Since we are sending a metadata request with specific topics instead of "asking for all topics", when `node != null` we will always see a `Errors.LEADER_NOT_AVAILABLE` on the per-partition error field, so this check should already be covered in line 1911 above.

guozhangwang replied:
Actually, nvm.

Just to clarify: `leaderFor` may return null either 1) the metadata cluster does not have this topic partition at all, or 2) the topic partition info exist, but its `leader` is null. For case 2) we should already have an error code and checked in line 1911 above already. But case 1) may still exist, for example, if the topic exist but with 4 partitions only and you are requesting to delete on that topic's partition 5.

guozhangwang reviewed (commented):
I think the jenkins failures are not relevant and should have been fixed. I left a few more minor comments, overall lgtm.

guozhangwang commented:
retest this please

asfgit commented:
FAILURE
 7968 tests run, 5 skipped, 2 failed.
--none--

asfgit commented:
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8076 tests run, 5 skipped, 0 failed.
--none--

ppatierno committed:
Saving references to errors and cluster from the metadata response

asfgit commented:
SUCCESS 
 8080 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8080 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
FAILURE
 7280 tests run, 1 skipped, 0 failed.
--none--

guozhangwang commented:
LGTM. Merged to trunk. Thanks @ppatierno

ppatierno commented:
@guozhangwang thanks to you and @cmccabe for your great reviewing work !
